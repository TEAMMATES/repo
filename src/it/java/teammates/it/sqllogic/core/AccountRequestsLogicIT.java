package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.UUID;

import org.testng.annotations.Test;

import teammates.common.datatransfer.AccountRequestStatus;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountRequestsLogic;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlentity.AccountRequest;

/**
 * SUT: {@link AccountRequestsLogic}.
 */
public class AccountRequestsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountRequestsLogic accountRequestsLogic = AccountRequestsLogic.inst();

    @Test
    public void testGetAccountRequest_nonExistentAccountRequest_returnsNull() {
        UUID id = UUID.randomUUID();
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertNull(actualAccountRequest);
    }

    @Test
    public void testGetAccountRequest_existingAccountRequest_getsSuccessfully() throws InvalidParametersException {
        AccountRequest expectedAccountRequest =
                new AccountRequest("test@gmail.com", "name", "institute", AccountRequestStatus.PENDING, "comments");
        UUID id = expectedAccountRequest.getId();
        accountRequestsLogic.createAccountRequest(expectedAccountRequest);
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertEquals(expectedAccountRequest, actualAccountRequest);
    }

    @Test
    public void testResetAccountRequest()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {

        ______TS("success: create account request and update registeredAt field");

        String name = "name lee";
        String email = "email@gmail.com";
        String institute = "institute";
        AccountRequestStatus status = AccountRequestStatus.PENDING;
        String comments = "comments";

        AccountRequest toReset = accountRequestsLogic.createAccountRequest(name, email, institute, status, comments);
        AccountRequestsDb accountRequestsDb = AccountRequestsDb.inst();

        toReset.setRegisteredAt(Instant.now());
        toReset = accountRequestsDb.getAccountRequest(email, institute);

        assertNotNull(toReset);
        assertNotNull(toReset.getRegisteredAt());

        ______TS("success: reset account request that already exists");

        AccountRequest resetted = accountRequestsLogic.resetAccountRequest(email, institute);

        assertNull(resetted.getRegisteredAt());

        ______TS("success: test delete account request");

        accountRequestsLogic.deleteAccountRequest(email, institute);

        assertNull(accountRequestsLogic.getAccountRequest(email, institute));

        ______TS("failure: reset account request that does not exist");

        assertThrows(EntityDoesNotExistException.class,
                () -> accountRequestsLogic.resetAccountRequest(name, institute));
    }
}
