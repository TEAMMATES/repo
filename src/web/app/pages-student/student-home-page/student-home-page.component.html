<ng-container *ngFor="let course of courses">
  <div class="card bg-light top-padded">
    <div class="card-header bg-primary text-white">
      <strong>[{{ course.id }}]: {{ course.name }}</strong>
      <span class="float-sm-right">
        <a routerLink="user={{ user }}&courseid={{ course.id }}" class="btn btn-light btn-sm" type="button">
          View team
        </a>
      </span>
    </div>

    <div class="table-responsive">
      <table class="table table-striped table-bordered margin-0" *ngIf="course.sessions.length">
        <thead>
        <tr>
          <th>Session Name</th>
          <th>Deadline</th>
          <th>Submissions</th>
          <th>Responses</th>
          <th>Action(s)</th>
        </tr>
        </thead>
        <tbody>
          <tr *ngFor="let session of course.sessions">
            <td>{{ session.name }}</td>
            <td>{{ session.deadline }}</td>
            <td>{{ session.isOpened | sessionSubmissionStatus: session.isWaitingToOpen : session.isSubmitted }}</td>
            <td>{{ session.isSessionPublished | sessionResponseStatus }}</td>
            <td>
              <a *ngIf="session.isSessionPublished" [href]="session.studentFeedbackResultsLink" class="btn btn-light btn-sm btn-left" type="button">
                View Responses
              </a>
              <a *ngIf="!session.isSessionPublished" class="btn btn-light btn-sm btn-left disabled" type="button">
                View Responses
              </a>
              <a *ngIf="session.isOpened && !session.isSubmitted" [href]="session.studentFeedbackResponseEditLink" class="btn btn-light btn-sm" type="button">
                Start Submission
              </a>
              <a *ngIf="session.isWaitingToOpen" class="btn btn-light btn-sm disabled" type="button">
                Start Submission
              </a>
              <a *ngIf="session.isOpened && session.isSubmitted" [href]="session.studentFeedbackResponseEditLink" class="btn btn-light btn-sm" type="button">
                Edit Submission
              </a>
              <a *ngIf="!session.isOpened && !session.isWaitingToOpen" [href]="session.studentFeedbackResponseEditLink" class="btn btn-light btn-sm" type="button">
                View Submission
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</ng-container>
