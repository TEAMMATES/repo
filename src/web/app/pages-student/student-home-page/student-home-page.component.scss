.margin-0 {
  margin: 0;
}

.alignment-buttons {
  width: 100px;
  height: 55px;
  margin-bottom: 5px;
}
