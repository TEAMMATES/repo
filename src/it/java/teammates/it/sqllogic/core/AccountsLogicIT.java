package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import org.testng.annotations.Test;

import teammates.common.datatransfer.NotificationStyle;
import teammates.common.datatransfer.NotificationTargetUser;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountsLogic;
import teammates.sqllogic.core.NotificationsLogic;
import teammates.storage.sqlapi.AccountsDb;
import teammates.storage.sqlentity.Account;
import teammates.storage.sqlentity.Notification;
import teammates.storage.sqlentity.ReadNotification;

/**
 * SUT: {@link AccountsLogic}.
 */
public class AccountsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountsLogic accountsLogic = AccountsLogic.inst();
    private NotificationsLogic notificationsLogic = NotificationsLogic.inst();

    private AccountsDb accountsDb = AccountsDb.inst();

    @Test
    public void testUpdateReadNotifications()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {
        ______TS("success: mark notification as read");
        Account account = new Account("google-id", "name", "email@teammates.com");
        Notification notification = new Notification(Instant.parse("2011-01-01T00:00:00Z"),
                Instant.parse("2099-01-01T00:00:00Z"), NotificationStyle.DANGER, NotificationTargetUser.GENERAL,
                "A deprecation note", "<p>Deprecation happens in three minutes</p>");
        accountsDb.createAccount(account);
        notificationsLogic.createNotification(notification);

        String googleId = account.getGoogleId();
        UUID notificationId = notification.getId();
        accountsLogic.updateReadNotifications(googleId, notificationId, notification.getEndTime());

        Account actualAccount = accountsDb.getAccountByGoogleId(googleId);
        List<ReadNotification> accountReadNotifications = actualAccount.getReadNotifications();
        assertEquals(1, accountReadNotifications.size());
        assertSame(actualAccount, accountReadNotifications.get(0).getAccount());
        assertSame(notification, accountReadNotifications.get(0).getNotification());
    }
}
