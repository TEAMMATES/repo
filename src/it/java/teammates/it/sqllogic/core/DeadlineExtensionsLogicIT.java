package teammates.it.sqllogic.core;

import java.time.Instant;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.HibernateUtil;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.DeadlineExtensionsLogic;
import teammates.storage.sqlentity.DeadlineExtension;
import teammates.storage.sqlentity.FeedbackSession;
import teammates.storage.sqlentity.Student;

/**
 * SUT: {@link DeadlineExtensionsLogic}.
 */
public class DeadlineExtensionsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private DeadlineExtensionsLogic deadlineExtensionsLogic = DeadlineExtensionsLogic.inst();
    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
        HibernateUtil.clearSession();
    }

    @Test
    public void testGetExtendedDeadline_extensionExists_success() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student1InCourse1");

        assert student != null;
        Instant extendedDeadlineForStudent = deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student);

        assertNotNull(extendedDeadlineForStudent);
        assertEquals(Instant.parse("2027-04-30T23:00:00Z"), extendedDeadlineForStudent);
    }

    @Test
    public void testGetExtendedDeadline_extensionDoesNotExist_null() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student2InCourse1");
        Instant extendedDeadlineForStudent = deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student);

        assertNull(extendedDeadlineForStudent);
    }

    @Test
    public void testUpdateDeadlineExtension()
            throws EntityDoesNotExistException, InvalidParametersException {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student1InCourse1");

        DeadlineExtension de = typicalDataBundle.deadlineExtensions.get("student1InCourse1Session1");

        Instant newEndTime = Instant.parse("2028-04-30T23:00:00Z");
        de.setEndTime(newEndTime);
        deadlineExtensionsLogic.updateDeadlineExtension(de);

        assertEquals(newEndTime, deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student));
    }

    @Test
    public void testUpdateDeadlineExtension_invalidParameters_originalUnchanged() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student1InCourse1");

        DeadlineExtension de = typicalDataBundle.deadlineExtensions.get("student1InCourse1Session1");
        Instant originalEndTime = de.getEndTime();

        Instant invalidEndTime = Instant.parse("2011-04-01T22:00:00Z");
        assert invalidEndTime.isBefore(feedbackSession.getStartTime());
        de.setEndTime(invalidEndTime);

        assertThrows(InvalidParametersException.class, () -> deadlineExtensionsLogic.updateDeadlineExtension(de));

        assertEquals(originalEndTime, deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student));
    }

}
