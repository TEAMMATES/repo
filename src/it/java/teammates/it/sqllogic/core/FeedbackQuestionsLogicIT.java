package teammates.it.sqllogic.core;

import java.util.ArrayList;
import java.util.List;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.FeedbackParticipantType;
import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.datatransfer.questions.FeedbackTextQuestionDetails;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.HibernateUtil;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.FeedbackQuestionsLogic;
import teammates.storage.sqlentity.FeedbackQuestion;
import teammates.storage.sqlentity.FeedbackSession;

/**
 * SUT: {@link FeedbackQuestionsLogic}.
 */
public class FeedbackQuestionsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private FeedbackQuestionsLogic fqLogic = FeedbackQuestionsLogic.inst();

    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
    }

    @Test
    public void testCreateFeedbackQuestion() throws InvalidParametersException {
        FeedbackSession fs = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        FeedbackTextQuestionDetails newQuestionDetails = new FeedbackTextQuestionDetails("New question text.");
        List<FeedbackParticipantType> showTos = new ArrayList<>();
        showTos.add(FeedbackParticipantType.INSTRUCTORS);
        FeedbackQuestion newQuestion = FeedbackQuestion.makeQuestion(fs, 6, "This is a new text question",
                FeedbackParticipantType.STUDENTS, FeedbackParticipantType.OWN_TEAM_MEMBERS, -100,
                showTos, showTos, showTos, newQuestionDetails);

        newQuestion = fqLogic.createFeedbackQuestion(newQuestion);

        FeedbackQuestion actualQuestion = fqLogic.getFeedbackQuestion(newQuestion.getId());

        verifyEquals(newQuestion, actualQuestion);
    }

    @Test
    public void testGetFeedbackQuestionsForSession() {
        FeedbackSession fs = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        FeedbackQuestion fq1 = typicalDataBundle.feedbackQuestions.get("qn1InSession1InCourse1");
        FeedbackQuestion fq2 = typicalDataBundle.feedbackQuestions.get("qn2InSession1InCourse1");
        FeedbackQuestion fq3 = typicalDataBundle.feedbackQuestions.get("qn3InSession1InCourse1");
        FeedbackQuestion fq4 = typicalDataBundle.feedbackQuestions.get("qn4InSession1InCourse1");
        FeedbackQuestion fq5 = typicalDataBundle.feedbackQuestions.get("qn5InSession1InCourse1");
        FeedbackQuestion fq6 = typicalDataBundle.feedbackQuestions.get("qn6InSession1InCourse1NoResponses");

        List<FeedbackQuestion> expectedQuestions = List.of(fq1, fq2, fq3, fq4, fq5, fq6);

        List<FeedbackQuestion> actualQuestions = fqLogic.getFeedbackQuestionsForSession(fs);

        assertEquals(expectedQuestions.size(), actualQuestions.size());
        assertTrue(expectedQuestions.containsAll(actualQuestions));
    }

}
