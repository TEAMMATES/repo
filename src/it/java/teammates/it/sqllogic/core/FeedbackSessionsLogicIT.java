package teammates.it.sqllogic.core;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.HibernateUtil;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.FeedbackSessionsLogic;
import teammates.storage.sqlentity.FeedbackSession;

/**
 * SUT: {@link FeedbackSessionsLogic}.
 */
public class FeedbackSessionsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private FeedbackSessionsLogic fsLogic = FeedbackSessionsLogic.inst();

    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
    }

    @Test
    public void testPublishFeedbackSession()
            throws InvalidParametersException, EntityDoesNotExistException {
        FeedbackSession unpublishedFs = typicalDataBundle.feedbackSessions.get("unpublishedSession1InTypicalCourse");

        FeedbackSession publishedFs1 = fsLogic.publishFeedbackSession(
                unpublishedFs.getName(), unpublishedFs.getCourse().getId());

        assertEquals(publishedFs1.getName(), unpublishedFs.getName());
        assertTrue(publishedFs1.isPublished());

        assertThrows(InvalidParametersException.class, () -> fsLogic.publishFeedbackSession(
                publishedFs1.getName(), publishedFs1.getCourse().getId()));
        assertThrows(EntityDoesNotExistException.class, () -> fsLogic.publishFeedbackSession(
                "non-existent name", unpublishedFs.getCourse().getId()));
        assertThrows(EntityDoesNotExistException.class, () -> fsLogic.publishFeedbackSession(
                unpublishedFs.getName(), "random-course-id"));
    }

    @Test
    public void testUnpublishFeedbackSession()
            throws InvalidParametersException, EntityDoesNotExistException {
        FeedbackSession publishedFs = typicalDataBundle.feedbackSessions.get("session1InCourse1");

        FeedbackSession unpublishedFs1 = fsLogic.unpublishFeedbackSession(
                publishedFs.getName(), publishedFs.getCourse().getId());

        assertEquals(unpublishedFs1.getName(), publishedFs.getName());
        assertFalse(unpublishedFs1.isPublished());

        assertThrows(InvalidParametersException.class, () -> fsLogic.unpublishFeedbackSession(
                unpublishedFs1.getName(), unpublishedFs1.getCourse().getId()));
        assertThrows(EntityDoesNotExistException.class, () -> fsLogic.unpublishFeedbackSession(
                "non-existent name", publishedFs.getCourse().getId()));
        assertThrows(EntityDoesNotExistException.class, () -> fsLogic.unpublishFeedbackSession(
                publishedFs.getName(), "random-course-id"));
    }

}
