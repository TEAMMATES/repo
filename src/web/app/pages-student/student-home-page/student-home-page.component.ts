import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import moment from 'moment-timezone';
import { FeedbackSessionsService } from '../../../services/feedback-sessions.service';

import { HttpRequestService } from '../../../services/http-request.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { TimezoneService } from '../../../services/timezone.service';
import {
  Course,
  Courses,
  FeedbackSession,
  FeedbackSessionPublishStatus,
  FeedbackSessions,
  FeedbackSessionSubmissionStatus,
  HasResponses,
} from '../../../types/api-output';
import { ErrorMessageOutput } from '../../error-message-output';

interface StudentCourse {
  course: Course;
  feedbackSessions: StudentSession[];
}

interface StudentSession {
  session: FeedbackSession;
  endTime: string;
  isOpened: boolean;
  isWaitingToOpen: boolean;
  isPublished: boolean;
  isSubmitted: boolean;
}

/**
 * Student home page.
 */
@Component({
  selector: 'tm-student-home-page',
  templateUrl: './student-home-page.component.html',
  styleUrls: ['./student-home-page.component.scss'],
})
export class StudentHomePageComponent implements OnInit {

  // Tooltip messages
  studentFeedbackSessionStatusPublished: string =
      'The responses for the session have been published and can now be viewed.';
  studentFeedbackSessionStatusNotPublished: string =
      'The responses for the session have not yet been published and cannot be viewed.';
  studentFeedbackSessionStatusAwaiting: string =
      'The session is not open for submission at this time. It is expected to open later.';
  studentFeedbackSessionStatusPending: string = 'The feedback session is yet to be completed by you.';
  studentFeedbackSessionStatusSubmitted: string = 'You have submitted your feedback for this session.';
  studentFeedbackSessionStatusClosed: string = ' The session is now closed for submissions.';

  user: string = '';

  hasEventualConsistencyMsg: boolean = false;
  courses: StudentCourse[] = [];

  constructor(private route: ActivatedRoute, private httpRequestService: HttpRequestService,
              private statusMessageService: StatusMessageService,
              private feedbackSessionsService: FeedbackSessionsService,
              private timezoneService: TimezoneService) {
    this.timezoneService.getTzVersion();
  }

  ngOnInit(): void {
    this.route.queryParams.subscribe((queryParams: any) => {
      this.user = queryParams.user;
      this.getStudentCourses();
    });
  }

  /**
   * Gets the courses and feedback sessions involving the student.
   */
  getStudentCourses(): void {
    const paramMap: { [key: string]: string } = {
      entitytype: 'student',
    };
    this.httpRequestService.get('/courses', paramMap).subscribe((resp: Courses) => {
      for (const course of resp.courses) {
        this.feedbackSessionsService.getFeedbackSessionsForStudent(course.courseId)
            .subscribe((fss: FeedbackSessions) => {
              const sortedFss: FeedbackSession[] = fss.feedbackSessions
                  .map((fs: FeedbackSession) => Object.assign({}, fs))
                  .sort((a: FeedbackSession, b: FeedbackSession) => (a.createdAtTimestamp >
                      b.createdAtTimestamp) ? 1 : (a.createdAtTimestamp === b.createdAtTimestamp) ?
                      ((a.submissionEndTimestamp > b.submissionEndTimestamp) ? 1 : -1) : -1);

              const studentSessions: StudentSession[] = [];
              for (const fs of sortedFss) {
                const endTime: string = moment(fs.submissionEndTimestamp).tz(fs.timeZone)
                    .format('ddd, DD MMM, YYYY, hh:mm A zz');
                const isOpened: boolean = this.isOpened(fs);
                const isWaitingToOpen: boolean = this.isWaitingToOpen(fs);
                const isPublished: boolean = this.isPublished(fs);
                this.feedbackSessionsService.hasStudentResponseForFeedbackSession(course.courseId,
                    fs.feedbackSessionName)
                    .subscribe((hasRes: HasResponses) => {
                      const isSubmitted: boolean = hasRes.hasResponses;
                      studentSessions.push(Object.assign({},
                          { endTime, isOpened, isWaitingToOpen, isPublished, isSubmitted, session: fs }));
                    });
              }

              this.courses.push(Object.assign({}, { course, feedbackSessions: studentSessions }));
              this.courses.sort((a: StudentCourse, b: StudentCourse) =>
                  (a.course.courseId > b.course.courseId) ? 1 : -1);
            });
      }
    }, (e: ErrorMessageOutput) => {
      this.statusMessageService.showErrorMessage(e.error.message);
    });
  }

  /**
   * Checks if feedback session is opened.
   */
  isOpened(fs: FeedbackSession): boolean {
    return fs.submissionStatus === FeedbackSessionSubmissionStatus.OPEN;
  }

  /**
   * Checks if feedback session is waiting to open.
   */
  isWaitingToOpen(fs: FeedbackSession): boolean {
    return fs.submissionStatus === FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN;
  }

  /**
   * Checks if feedback session is published.
   */
  isPublished(fs: FeedbackSession): boolean {
    return fs.publishStatus === FeedbackSessionPublishStatus.PUBLISHED;
  }

  /**
   * Gets the tooltip message for the submission status.
   */
  getSubmissionStatusTooltip(session: StudentSession): string {
    let msg: string = '';

    if (session.isWaitingToOpen) {
      msg += this.studentFeedbackSessionStatusAwaiting;
    } else if (session.isSubmitted) {
      msg += this.studentFeedbackSessionStatusSubmitted;
    } else {
      msg += this.studentFeedbackSessionStatusPending;
    }
    if (!session.isOpened && !session.isWaitingToOpen) {
      msg += this.studentFeedbackSessionStatusClosed;
    }
    return msg;
  }

  /**
   * Gets the tooltip message for the response status.
   */
  getResponseStatusTooltip(isPublished: boolean): string {
    if (isPublished) {
      return this.studentFeedbackSessionStatusPublished;
    }
    return this.studentFeedbackSessionStatusNotPublished;
  }
}
