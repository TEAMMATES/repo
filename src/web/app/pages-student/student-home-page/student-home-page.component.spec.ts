import { HttpClientTestingModule } from '@angular/common/http/testing';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { StudentHomePageComponent } from './student-home-page.component';

import { ResponseStatusPipe } from '../../pipes/session-response-status.pipe';
import { SubmissionStatusPipe } from '../../pipes/session-submission-status.pipe';

describe('StudentHomePageComponent', () => {
  let component: StudentHomePageComponent;
  let fixture: ComponentFixture<StudentHomePageComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        StudentHomePageComponent,
        ResponseStatusPipe,
        SubmissionStatusPipe,
      ],
      imports: [
        HttpClientTestingModule,
        NgbModule,
        RouterTestingModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(StudentHomePageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with null fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with blank fields', () => {
    const studentName: string = '';
    const studentCourse: any = {
      course: {},
      feedbackSessions: [],
    };
    component.user = studentName;
    component.courses = [studentCourse];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with populated fields', () => {
    const studentName: string = 'John Doe';
    const studentCourse: any = {
      course: {
        id: 'CS2103',
        name: 'Software Engineering',
      },
      feedbackSessions: [],
    };
    const sessionInfoMap: any = {
      endTime: '1200',
      isOpened: true,
      isWaitingToOpen: true,
      isPublished: true,
      isSubmitted: true,
    };

    component.user = studentName;
    component.courses = [studentCourse];
    component.sessionsInfoMap.set('Session', sessionInfoMap);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
