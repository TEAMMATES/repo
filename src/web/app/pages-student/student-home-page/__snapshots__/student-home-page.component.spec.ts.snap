// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentHomePageComponent should snap with blank fields 1`] = `
<tm-student-home-page
  courses={[Function Array]}
  hasEventualConsistencyMsg="false"
  httpRequestService={[Function HttpRequestService]}
  recentlyJoinedCourseId=""
  route={[Function ActivatedRoute]}
  sessionsInfoMap={[Function Map]}
  statusMessageService={[Function StatusMessageService]}
  studentFeedbackSessionStatusAwaiting={[Function String]}
  studentFeedbackSessionStatusClosed={[Function String]}
  studentFeedbackSessionStatusNotPublished={[Function String]}
  studentFeedbackSessionStatusPending={[Function String]}
  studentFeedbackSessionStatusPublished={[Function String]}
  studentFeedbackSessionStatusSubmitted={[Function String]}
  user=""
>
  <div
    class="card bg-light top-padded"
  >
    
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        []: 
      </strong>
      <span
        class="float-sm-right"
      >
        <a
          class="btn btn-light btn-sm button"
          href="/course?user="
          ng-reflect-ngb-tooltip="View and edit information rega"
          ng-reflect-query-params="[object Object]"
          ng-reflect-router-link="../course"
          ngbtooltip="View and edit information regarding your team"
          routerlink="../course"
        >
           View team 
        </a>
      </span>
    </div>
    <div
      class="table-responsive"
    >
      
      <table
        class="table table-striped table-bordered margin-0"
      >
        <thead>
          <tr>
            <th>
              Session Name
            </th>
            <th>
              Deadline
            </th>
            <th>
              Submissions
            </th>
            <th>
              Responses
            </th>
            <th>
              Action(s)
            </th>
          </tr>
        </thead>
        <tbody>
          
        </tbody>
      </table>
      
    </div>
  </div>
</tm-student-home-page>
`;

exports[`StudentHomePageComponent should snap with null fields 1`] = `
<tm-student-home-page
  courses={[Function Array]}
  hasEventualConsistencyMsg="false"
  httpRequestService={[Function HttpRequestService]}
  recentlyJoinedCourseId=""
  route={[Function ActivatedRoute]}
  sessionsInfoMap={[Function Map]}
  statusMessageService={[Function StatusMessageService]}
  studentFeedbackSessionStatusAwaiting={[Function String]}
  studentFeedbackSessionStatusClosed={[Function String]}
  studentFeedbackSessionStatusNotPublished={[Function String]}
  studentFeedbackSessionStatusPending={[Function String]}
  studentFeedbackSessionStatusPublished={[Function String]}
  studentFeedbackSessionStatusSubmitted={[Function String]}
  user="undefined"
>
  
</tm-student-home-page>
`;

exports[`StudentHomePageComponent should snap with populated fields 1`] = `
<tm-student-home-page
  courses={[Function Array]}
  hasEventualConsistencyMsg="false"
  httpRequestService={[Function HttpRequestService]}
  recentlyJoinedCourseId=""
  route={[Function ActivatedRoute]}
  sessionsInfoMap={[Function Map]}
  statusMessageService={[Function StatusMessageService]}
  studentFeedbackSessionStatusAwaiting={[Function String]}
  studentFeedbackSessionStatusClosed={[Function String]}
  studentFeedbackSessionStatusNotPublished={[Function String]}
  studentFeedbackSessionStatusPending={[Function String]}
  studentFeedbackSessionStatusPublished={[Function String]}
  studentFeedbackSessionStatusSubmitted={[Function String]}
  user={[Function String]}
>
  <div
    class="card bg-light top-padded"
  >
    
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        [CS2103]: Software Engineering
      </strong>
      <span
        class="float-sm-right"
      >
        <a
          class="btn btn-light btn-sm button"
          href="/course?user=John%20Doe&courseid=CS2103"
          ng-reflect-ngb-tooltip="View and edit information rega"
          ng-reflect-query-params="[object Object]"
          ng-reflect-router-link="../course"
          ngbtooltip="View and edit information regarding your team"
          routerlink="../course"
        >
           View team 
        </a>
      </span>
    </div>
    <div
      class="table-responsive"
    >
      
      <table
        class="table table-striped table-bordered margin-0"
      >
        <thead>
          <tr>
            <th>
              Session Name
            </th>
            <th>
              Deadline
            </th>
            <th>
              Submissions
            </th>
            <th>
              Responses
            </th>
            <th>
              Action(s)
            </th>
          </tr>
        </thead>
        <tbody>
          
        </tbody>
      </table>
      
    </div>
  </div>
</tm-student-home-page>
`;
